Resyncs RichText mobile components with web counterparts.

SergioEstevao opened the pull request:
## Description
<!-- Please describe what you have changed or added -->
This PR updates the RN/mobile RichText components to be more in sync with their web counterparts.

The web counterparts have been updated in the last 4 months and now they have more clear separation between the rich text component that is block aware and the rich text component that is only worried about HTML/format presentation/edition.

This bring the same separation for the RN mobile components.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

This can be tested using this GB-mobile PR:https://github.com/wordpress-mobile/gutenberg-mobile/pull/1437
## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

RN Mobile
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

SergioEstevao committed:
Resyncs RichText mobile components with web counterparts.

SergioEstevao committed:
Remove outdated test.

SergioEstevao committed:
Remove unused references.

etoledom commented:
It looks like a big part of `packages/block-editor/src/components/rich-text/index.native.js` is a direct copy of the `index.js` counterpart. 

Do you think it would be worth to have a `common.js` file to share that code?

SergioEstevao committed:
Merge branch 'master' into rnmobile/rich_text_sync

SergioEstevao commented:
> It looks like a big part of packages/block-editor/src/components/rich-text/index.native.js is a direct copy of the index.js counterpart.

> Do you think it would be worth to have a common.js file to share that code?

Definitely but for this PR I just wanted to update first, we can do a follow up PR were we see parts that can abstracted to separate methods and files. 

The main changes are related to the toolbars/UI, paste behaviour when copying files/images and some extra props around block selection/focus.

ellatrix commented:
It would definitely be good if `packages/block-editor/src/components/rich-text/index.js` could be reused entirely.

koke commented:
I'm having a hard time reviewing this, and GitHub isn't helping. I even tried a 3-way diff but it's not super useful.

![Screen Shot 2019-10-16 at 09 36 16](https://user-images.githubusercontent.com/8739/66897995-868fa700-eff8-11e9-8d50-d63ee42e10de.png)

Can this be done in smaller chunks, or can you add more details documenting what has changed? I see that the block-editor one is almost identical to the web version, but the one in components has a bunch of changes that are harder to parse, since it's still very different from the web:

```
koke:gutenberg/ (rnmobile/rich_text_sync*) $ diff -U0 component-rich-text_HEAD.js component-rich-text_WEB.js | wc -l
    1445
koke:gutenberg/ (rnmobile/rich_text_sync*) $ diff -U0 component-rich-text_HEAD.js component-rich-text_BASE.js | wc -l
     297
koke:gutenberg/ (rnmobile/rich_text_sync*) $ diff -U0 component-rich-text_BASE.js component-rich-text_WEB.js | wc -l
    1631
```

SergioEstevao commented:
It's a a bit hard to change step by step now, because of the amount of
changes that happened in the last months on the web counterpart.

I basically copy pasted all the web file in the block library to the native
file and then added the changes needed for native and adapted the component
to the new interface.



On Wed, 16 Oct 2019 at 08:45, Jorge Bernal <notifications@github.com> wrote:

> I'm having a hard time reviewing this, and GitHub isn't helping. I even
> tried a 3-way diff but it's not super useful.
>
> [image: Screen Shot 2019-10-16 at 09 36 16]
> <https://user-images.githubusercontent.com/8739/66897995-868fa700-eff8-11e9-8d50-d63ee42e10de.png>
>
> Can this be done in smaller chunks, or can you add more details
> documenting what has changed? I see that the block-editor one is almost
> identical to the web version, but the one in components has a bunch of
> changes that are harder to parse, since it's still very different from the
> web:
>
> koke:gutenberg/ (rnmobile/rich_text_sync*) $ diff -U0 component-rich-text_HEAD.js component-rich-text_WEB.js | wc -l
>     1445
> koke:gutenberg/ (rnmobile/rich_text_sync*) $ diff -U0 component-rich-text_HEAD.js component-rich-text_BASE.js | wc -l
>      297
> koke:gutenberg/ (rnmobile/rich_text_sync*) $ diff -U0 component-rich-text_BASE.js component-rich-text_WEB.js | wc -l
>     1631
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/17897?email_source=notifications&email_token=AAE7CUN3HB4SSWRZT5MVYG3QO3BCDA5CNFSM4I7XEBWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBLPKOQ#issuecomment-542569786>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE7CUIIXPF33EFZKXERH2DQO3BCDANCNFSM4I7XEBWA>
> .
>

SergioEstevao commented:
@koke in a way it will be easier to compare the `packages/block-editor/src/components/rich-text/index.native.js ` file to it's web version to see the changes that are needed for the native side to work.

ellatrix commented:
Might it be perhaps easier to follow if some code is immediately reused. I'm not sure what the differences are at this stage. It would be good to have a list of them, and if you decide to go for a separate native file still, to provide comment on the parts that diverge.

For example, I see some diverging paste logic, which might also be useful for web. If you decide to edit the web code, _that_ I could help review.

SergioEstevao committed:
Add platform component

SergioEstevao committed:
Add components depending of platform.

Only add specific components if we are on the web implementation.

SergioEstevao committed:
Abstract paste of files for RN and web

Makes the code for pasting image more abstract in the paste method and
implement specific translation to HTML depending of the platform.

SergioEstevao committed:
Compose extra attributes/props on select/dispatch only if mobile.

SergioEstevao committed:
Remove RN index file for RichText Wrapper.

Moved all the specific code to the standard index file, so this file
is no longer needed.

SergioEstevao commented:
@koke, @ellatrix, @etoledom I updated the PR in order to have a single `packages/block-editor/src/components/rich-text/index.native.js` file that has some platform checks for the specific native changes/limitations.

Does this looks better?

koke commented on packages/block-editor/src/components/rich-text/file-paste-handler.native.js:
```diff
@@ -0,0 +1,20 @@
+ */
+import { createBlobURL } from '@wordpress/blob';
+
+export function filePasteHandler(images, html) {
```
I'm trying to understand the differences with the web on this one, and what I'm seeing:

- The native version adds a 'wp-image-{uploadId}' class. This seems related to the media upload in the apps, but I'm not sure if it's actually used, since pasting an image on RichText would end up transforming to an Image block, and that would request a media import anyway.
- The web version receives a [File](https://developer.mozilla.org/en-US/docs/Web/API/File) object, while native receives paths to the files. I imagine we'll see more of this distinction elsewhere and I wonder if we should override `@wordpress/blob` instead so that our "blob" URLs are actually the passed file URLs (i.e. `createBlobURL = url => url`).
- The web only processes files when there was no html pasted, but I think we could do the same.

Am I missing extra context, or could these two merge into a single implementation?

SergioEstevao replied:
> The native version adds a 'wp-image-{uploadId}' class. This seems related to the media upload in the apps, but I'm not sure if it's actually used, since pasting an image on RichText would end up transforming to an Image block, and that would request a media import anyway.

I think the reason the upload happens is because we set this wp-image-uploadId or else the image block will just refer to a local file. No?

koke replied:
I'm not super familiar with the whole process, but what I see on the image block is that it will `requestMediaImport` for any image with a `file:` URL. https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/image/edit.native.js#L107

SergioEstevao replied:
> The web only processes files when there was no html pasted, but I think we could do the same.

Do you know the rational behind this? Why can't we have both? 

If we want to do this on the native side too, we will need to check for empty string because our paste code sends an empty HTML string when no HTML is available: https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/react-native-aztec/ios/RNTAztecView/RCTAztecView.swift#L245

koke replied:
It looks like it was a fix for one of those “fun” bugs https://github.com/WordPress/gutenberg/issues/4854

koke replied:
> If we want to do this on the native side too, we will need to check for empty string

I think an empty string would evaluate as false, so the current check should work

koke commented on packages/blocks/src/api/index.native.js:
```diff
@@ -2,6 +2,8 @@ export {
 	createBlock,
 	switchToBlockType,
+	getBlockTransforms,
+	findTransform,
```
There doesn't seem to be that many differences in this file anymore. Have you tried removing the native variant?

koke commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -509,13 +520,27 @@ const RichTextContainer = compose( [
 		}
 
+		let extraProps = {}
+		if ( Platform.OS !== 'web' ) {
```
These two props don't seem to conflict with the web stuff, so maybe we can skip the check and always inject them.  Maybe we could have something like a `__native` prefix to indicate better that they are for native-specific behavior.

SergioEstevao replied:
Will the check for platform not make it obvious enough? I could change the code to `Platform.OS == `native` now that we have our Platform component in.

koke commented:
ESLint is showing a bunch of issues:

```
✖ 231 problems (81 errors, 150 warnings)
79 errors and 1 warning potentially fixable with the `--fix` option.
```

SergioEstevao committed:
Remove API index native file that is no longer needed.

SergioEstevao commented on packages/element/src/index.js:
```diff
@@ -1,5 +1,6 @@
 export * from './react';
 export * from './react-platform';
 export * from './utils';
+export { default as Platform } from './platform'; 
```
That is the plan, I've been cherry picking from that branch to make the merge easier later on.

SergioEstevao commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -482,19 +481,31 @@ class RichTextWrapper extends Component {
+			return {
+				clientId,
+				blockIsSelected: ownProps.isSelected !== undefined ? ownProps.isSelected : isSelected,
+				onCaretVerticalPositionChange,
```
I'm also not sure why we needed those props, maybe @hypest or @Tug remember the reasons?

hypest replied:
I see https://github.com/WordPress/gutenberg/pull/13323 that introduced the `onCaretVerticalPositionChange`

Tug replied:
And https://github.com/WordPress/gutenberg/pull/15878 introduced `blockIsSelected` to fix a focus issue.

Tug replied:
We should definitely look into it in a followup PR, there might be "cleaner" ways to address those issues

koke commented on packages/rich-text/src/component/index.native.js:
```diff
@@ -346,93 +265,62 @@ export class RichText extends Component {
-		}
-
+	handleDelete( event ) {
 		const keyCode = BACKSPACE; // TODO : should we differentiate BACKSPACE and DELETE?
```
I don't think Aztec is currently capturing `DELETE` but it would be a nice thing to have as well. I don't see a way to inject the `keycode` prop into the event from the native side, but we could have `AztecView` do that if we want to keep the support for `DELETE` here.

SergioEstevao replied:
I was thinking to change the code on our AztecView to send keycodes the same way as the web, but we will only send up DELETE, ENTER and BACKSPACE for now.
 But I was considering that for another PR more specific to AztecView.

koke replied:
👍 I think it's fine to leave it for later

koke reviewed (approved):
This is looking much better. I left some comments on the code, and I think at least another set of more experienced eyes would be good before merging. Testing seemed fine, although I noticed that when pasting multiple paragraphs, the cursor ended up at the beginning of the last pasted block.

Besides ESLint, there's also the Build Artifacts job failing, I'm guessing because there are some autogenerated docs that aren't committed. Running the pre-commit hooks locally should help there.

I can do another round of testing when others approve the code changes.

koke commented:
Note that I only tested iOS, since I'm having some issues lately with my Android setup 😞

SergioEstevao committed:
Clean up lint errors in file-paste-handler.

SergioEstevao committed:
Fix lint errors.

SergioEstevao committed:
Implement stub remove browser shortcuts for RN

ellatrix commented on packages/block-editor/src/components/rich-text/file-paste-handler.native.js:
```diff
@@ -0,0 +1,15 @@
+
```
What's stopping us from reusing this for web? What's the reason for this `Number.MAX_SAFE_INTEGER` stuff? Could that be commented on?

ellatrix commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -429,7 +426,7 @@ class RichTextWrapper extends Component {
 							</BlockFormatControls>
 						) }
-						{ isSelected && inlineToolbar && hasFormats && (
+						{ Platform.OS === 'web' && isSelected && inlineToolbar && hasFormats && (
```
Or extract it in a separate component?

ellatrix commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -440,25 +437,27 @@ class RichTextWrapper extends Component {
-							}
-						</Autocomplete>
+						{ Platform.OS === 'web' && isSelected && <RemoveBrowserShortcuts /> }
+						{ Platform.OS === 'web' && (
```
That sounds good!

SergioEstevao committed:
Implement autocomplete stub for RN.

ellatrix commented on packages/rich-text/src/component/index.js:
```diff
@@ -235,7 +235,7 @@ class RichText extends Component {
 			// Only process file if no HTML is present.
 			// Note: a pasted file may have the URL as plain text.
-			const image = find( [ ...items, ...files ], ( { type } ) =>
+			const image = filter( [ ...items, ...files ], ( { type } ) =>
```
Filter implies that an array would be returned, right? So this needs to be `images`?

ellatrix reviewed (changes requested):
Looks really good!

I'm just getting an error when pasting an image into rich text on the web.

```
index.js:16 Uncaught TypeError: Failed to execute 'createObjectURL' on 'URL': No function was found that matched the signature provided.
    at createBlobURL (index.js:16)
    at file-paste-handler.js:12
    at Array.forEach (<anonymous>)
    at filePasteHandler (file-paste-handler.js:9)
    at RichTextWrapper.onPaste (index.js:137)
    at RichText.onPaste (index.js:242)
    at HTMLUnknownElement.callCallback (react-dom.8b3dda97.js:341)
    at Object.invokeGuardedCallbackDev (react-dom.8b3dda97.js:391)
    at invokeGuardedCallback (react-dom.8b3dda97.js:448)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.8b3dda97.js:462)
```

SergioEstevao committed:
Refactor toolbar presentation to a method.

SergioEstevao committed:
Merge remote-tracking branch 'origin/master' into rnmobile/rich_text_sync

# Conflicts:
#	packages/element/src/index.js
#	packages/element/src/platform.js

SergioEstevao committed:
Remove no longer needed platform file.

SergioEstevao committed:
Consolidate the file paste handler in a single implementation.

Created a stub for createBlobURL for native that simple returns the
original URL.

SergioEstevao committed:
Change the text for platform to make it explicit it's native only.

ellatrix commented:
Tested again. I now get two images if I paste an image:

<img width="787" alt="Screenshot 2019-10-25 at 11 40 01" src="https://user-images.githubusercontent.com/4710635/67560980-462cd900-f71c-11e9-8dd4-3ce90830d231.png">

SergioEstevao commented:
@koke and @ellatrix I think I addressed all the issues raised, do you mind give it another look?

ellatrix commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -327,6 +324,29 @@ class RichTextWrapper extends Component {
 	}
 
+	renderFormatToolbar( inline ) {
+		if ( inline && Platform.OS === 'web ' ) {
```
Is there any way we can split this in a separate file and use the current `.native` system?

ellatrix commented on packages/rich-text/src/component/index.js:
```diff
@@ -260,7 +260,7 @@ class RichText extends Component {
 			// Only process file if no HTML is present.
 			// Note: a pasted file may have the URL as plain text.
-			const image = find( [ ...items, ...files ], ( { type } ) =>
+			const images = filter( [ ...items, ...files ], ( { type } ) =>
```
This array merging is probably the cause of the duplicated images. We should make sure there are no duplicates, or use one or the other (if one doesn't exist).

ellatrix replied:
I think the safest is to still use both, but discard any duplicates.

ellatrix replied:
Looks like you'll have to compare the `name` and `size`.

SergioEstevao replied:
@ellatrix I was thinking on trimming the array to size 1, but if you can clear the duplicates it will be even better! Feel free to push to the PR :)

ellatrix replied:
Done! :)

ellatrix commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -482,19 +487,30 @@ class RichTextWrapper extends Component {
 	withInstanceId,
-	withBlockEditContext( ( { clientId } ) => ( { clientId } ) ),
+	withBlockEditContext( ( { clientId, onCaretVerticalPositionChange, isSelected }, ownProps ) => {
+		if ( Platform.OS === 'web' ) {
```
An idea to keep the current `.native` system: create the higher order component in a separate fire where it returns the same component for web, and import it here. Native could have two `withBlockEditContext` higher order components.

ellatrix commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -509,13 +525,27 @@ const RichTextContainer = compose( [
+			const block = clientId && __unstableGetBlockWithoutInnerBlocks( clientId );
+			const shouldBlurOnUnmount = block && isSelected && isUnmodifiedDefaultBlock( block );
+			extraProps = {
+				blockIsSelected,
```
Does `blockIsSelected` need to be set again? It was already set in the previous higher order component, no?

SergioEstevao replied:
@hypest do you know why we need to do it twice or this is a duplicate?

hypest replied:
Looks like superfluous. The second HOC is not modifying the value at all so, no need to extract and return it again.

ellatrix committed:
Remove duplicate files

ellatrix committed:
Include type in file comparison

SergioEstevao commented on packages/block-editor/src/components/rich-text/file-paste-handler.js:
```diff
@@ -3,16 +3,18 @@
+	window.console.log( 'Received items:\n\n', files );
+
+	return files
+		.filter( ( { type } ) => /^image\/(?:jpe?g|png|gif)$/.test( type ) )
```
I don't think this will for the native files, they are file URL that point to the local file system on the device.
Can you do the filter before we get to this point?

ellatrix replied:
Oh, I see. What's the type in that case? It would be good to keep this image logic together.

ellatrix replied:
In other words, what does `files` look like for native?

ellatrix replied:
I see this comes straight from the event: `const { pastedText, pastedHtml, files } = event.nativeEvent;`.

ellatrix commented on packages/rich-text/src/component/index.js:
```diff
@@ -258,18 +258,32 @@ class RichText extends Component {
-			// Only process file if no HTML is present.
-			// Note: a pasted file may have the URL as plain text.
-			const image = find( [ ...items, ...files ], ( { type } ) =>
-				/^image\/(?:jpe?g|png|gif)$/.test( type )
```
Moved this check to the RichText for block implementation (which only handles images), so we generally pass any files.

ellatrix committed:
Forgot to rename for native file

ellatrix committed:
Fix filePasteHandler for native

ellatrix reviewed (approved):
On the web side, this looks good now.

ellatrix committed:
Move logging back

ellatrix committed:
Restore comment on logging

hypest commented:
Tried copying text from Chrome on Android (https://www.lipsum.com/) and pasting results in a redscreen with: `undefined is not an object (evaluating 'files.length')` at `onPaste`, `index.js:139:13`

SergioEstevao committed:
Add check for files existence.

SergioEstevao committed:
Refactor format-toolbar code to use split web/native files

SergioEstevao committed:
Remove prop duplication.

SergioEstevao committed:
Fix getAnchorRect call

SergioEstevao committed:
Remove unnecessary const

SergioEstevao committed:
Merge branch 'master' into rnmobile/rich_text_sync

SergioEstevao committed:
Sync fix for list removal of first empty line

SergioEstevao committed:
Fix RN build after merge with master

SergioEstevao committed:
Sync with web counterpart.

ellatrix commented:
I checked this again, and everything still looks and works good for web. Feel free to merge.

SergioEstevao committed:
Only change selection after new formats are set.
